packages/fxa-auth-server/scripts/paypal-processor.ts

```diff
@@ -53,17 +90,43 @@ export async function init() {
   );
   const statsd = Container.get(StatsD);
   statsd.increment('paypal-processor.startup');
-  await processor.processInvoices();
+
+  const timer = initTimer();
+
+  if (useLock) {
+    try {
+      const redis = new Redis(config.redis);
+      const redlock = new Redlock([redis], { retryCount: 1 });
+      lock = await redlock.acquire([program.lockName], lockDuration);
+    } catch (err) {
+      throw new Error(`Cannot acquire lock to run: ${err.message}`);
+    }
+  }
+
+  for await (const _ of processor.processInvoices()) {
+    if (useLock && timer.elapsed() > Math.floor(lockDuration / 2)) {
+      await lock.extend(timer.elapsed());
+      timer.reset();
+    }
+  }
+
   statsd.increment('paypal-processor.shutdown');
   await promisify(statsd.close).bind(statsd)();
   return 0;
 }
 
 if (require.main === module) {
+  let exitStatus = 1;
   init()
+    .then((result) => {
+      exitStatus = result;
+    })
     .catch((err) => {
       console.error(err);
-      process.exit(1);
+      exitStatus = 1;
     })
-    .then((result) => process.exit(result));
+    .finally(() => {
+      lock?.release();
```

Comment:
Is there a chance that the `.release()` API can throw an error?
Does this need to be wrapped in a `try..catch` for extra safety?

I guess the risk is that we make it to the `finally()` block and the lock release throws an error and our `process.exit()` never gets called (or would terminate with a 0 status code). Although maybe the thrown error would break elsewhere and it's a non-issue.

Reply:
Thanks.  I moved the `.release()` call into `init` so that it's in a `try` block.